lineapy/plugins/ray_pipeline_writer.py

```diff
@@ -0,0 +1,130 @@
+        file.write_text(full_code)
+        logger.info(f"Generated DAG file: {file}")
+
+    def _write_operators(
+        self,
+        dag_flavor: RayDagFlavor,
+    ) -> str:
+        """
+        Returns a code block containing all the operators for a Ray DAG.
+        """
+
+        DAG_TEMPLATE = load_plugin_template("ray_dag.jinja")
+
+        if dag_flavor == RayDagFlavor.TaskPerSession:
+            task_breakdown = DagTaskBreakdown.TaskPerSession
+        elif dag_flavor == RayDagFlavor.TaskPerArtifact:
+            task_breakdown = DagTaskBreakdown.TaskPerArtifact
+
+        # Get task definitions based on dag_flavor
+        task_defs, task_graph = get_task_graph(
+            self.artifact_collection,
+            pipeline_name=self.pipeline_name,
+            task_breakdown=task_breakdown,
+        )
+
+        rendered_task_defs = self.get_rendered_task_definitions(task_defs)
+
+        input_parameters_dict: Dict[str, Any] = {}
+        for parameter_name, input_spec in super().get_pipeline_args().items():
+            input_parameters_dict[parameter_name] = input_spec.value
+
+        full_code = DAG_TEMPLATE.render(
+            DAG_NAME=self.pipeline_name,
+            MODULE_NAME=self.pipeline_name + "_module",
+            init_args=self.dag_config.get("init_args", ""),
+            task_definitions=rendered_task_defs,
+            tasks=task_defs,
+            dag_params=input_parameters_dict,
+            sink_tasks=task_graph.sink_nodes,
+        )
+
+        return prettify(full_code)
+
+    @property
+    def docker_template_name(self) -> str:
+        return "ray_dockerfile.jinja"
+
+    def get_rendered_task_definitions(
+        self,
+        task_defs: Dict[str, TaskDefinition],
+    ) -> List[str]:
+        """
+        Returns rendered tasks for the pipeline tasks along with a dictionary to lookup
+        previous task outputs.
+
+        The returned dictionary is used by the DAG to connect the right input files to
+        output files for inter task communication.
+        """
+        TASK_FUNCTION_TEMPLATE = load_plugin_template(
+            "task/task_function.jinja"
+        )
+        rendered_task_defs: List[str] = []
+
+        for task_name, task_def in task_defs.items():
+
+            input_vars = (
+                task_def.user_input_variables + task_def.loaded_input_variables
+            )
+
+            task_def_rendered = TASK_FUNCTION_TEMPLATE.render(
+                function_decorator=f"@ray.remote(num_returns={len(task_def.return_vars)})",
+                function_name=task_name,
+                user_input_variables=", ".join(input_vars),
+                typing_blocks=task_def.typing_blocks,
+                pre_call_block="",
+                call_block=task_def.call_block,
+                post_call_block=task_def.post_call_block,
+                return_block=f"return {', '.join(task_def.return_vars)}",
+            )
+            rendered_task_defs.append(task_def_rendered)
```

Comment:
recommend moving this entire thing to base writer. its almost exactly same in all the writers and is a needless lift. we can pull out the task_loading_block piece outside this since it doesnt look like it is dependent on anything else in this function. This function also has no "self" dependencies and could become a pure function.

Reply:
Ok, putting this PR on hold for Argo to go first and then I will make this change.

Reply:
Addressing in separate PR